### PR TITLE
Drop support for Python 3.6

### PR DESCRIPTION
**Pull request by hramezani**

### CODE REVIEW

File: src/urllib3/_collections.py [hunk 1]
```diff
@@ -268,7 +245,7 @@ class HTTPHeaderDict(MutableMapping[str, str]):
 
     def __init__(self, headers: Optional[ValidHTTPHeaderSource] = None, **kwargs: str):
         super().__init__()
-        self._container = _ordered_dict()
+        self._container = dict()  # 'dict' is insert-ordered in Python 3.7+
```
Comment (sethmlarson) [inline src/urllib3/_collections.py:248]: nit: Can do an empty dict literal here

**Review by sethmlarson — APPROVED** [review]
LGTM, thanks for this. One nit for you:

**Comment by pquentin** [comment]
@sethmlarson can you please remove the PyPy build from the required checks? The name has changed to 3.7 and it's unreliable enough to not be required.

**Comment by V1NAY8** [comment]
@hramezani  Just did some digging Should we also update these comments or its underlying code?
https://github.com/urllib3/urllib3/blob/21945b84b0707a3143b09ade5c59e8dbf6a69a0f/src/urllib3/connection.py#L486

https://github.com/urllib3/urllib3/blob/21945b84b0707a3143b09ade5c59e8dbf6a69a0f/src/urllib3/util/ssl_.py#L195

https://github.com/urllib3/urllib3/blob/21945b84b0707a3143b09ade5c59e8dbf6a69a0f/src/urllib3/util/ssl_.py#L383

https://github.com/urllib3/urllib3/blob/21945b84b0707a3143b09ade5c59e8dbf6a69a0f/src/urllib3/util/ssl_match_hostname.py#L1

https://github.com/urllib3/urllib3/blob/21945b84b0707a3143b09ade5c59e8dbf6a69a0f/src/urllib3/util/ssl_match_hostname.py#L12

https://github.com/urllib3/urllib3/blob/21945b84b0707a3143b09ade5c59e8dbf6a69a0f/src/urllib3/contrib/pyopenssl.py#L286

https://github.com/urllib3/urllib3/blob/21945b84b0707a3143b09ade5c59e8dbf6a69a0f/src/urllib3/contrib/securetransport.py#L535

**Comment by hramezani** [comment]
Thanks @V1NAY8 for checking.
I think we can remove `(requires Python3.4+) `.
But, I think we can keep `# Copy-pasted from Python 3.5 source code `.

What do you think @sethmlarson @pquentin?

**Comment by pquentin** [comment]
@hramezani Agreed.

Regarding https://github.com/urllib3/urllib3/blob/main/src/urllib3/util/ssl_match_hostname.py, it's from Python 3.5 now, not 3.3, see https://github.com/urllib3/urllib3/commit/f620d997134708b09560ca5797aa79a59a2ef4c0. Also, it was modified to remove commonName support, which should be mentioned.

**Comment by hramezani** [comment]
Anything to do here? @sethmlarson Have you checked @pquentin [comment](https://github.com/urllib3/urllib3/pull/2336#issuecomment-885527422)?

**Review by sethmlarson — APPROVED** [review]
Thanks all for carrying this so far! I removed `pypy3` from the required checks on GitHub for the `main` branch, is there anything else I'm missing @pquentin and @hramezani?

File: test/test_ssltransport.py [hunk 2]
```diff
@@ -367,13 +367,10 @@ def test_wrong_sni_hint(self):
         with self.client_context.wrap_socket(
             sock, server_hostname="localhost"
         ) as proxy_sock:
-            with pytest.raises(Exception) as e:
+            with pytest.raises(ssl.SSLError):
```
Comment (pquentin) [inline test/test_ssltransport.py:370]: Can we catch https://docs.python.org/3/library/ssl.html#ssl.SSLCertVerificationError instead?

**Review by sethmlarson — APPROVED** [review]

**Review by pquentin — APPROVED** [review]

**Pull request merged**